File: pyserver/app/widgets/line_chart.py
from .base import BaseWidget
from PIL import Image, ImageDraw
from typing import Any

class LineChartWidget(BaseWidget):
    def draw(self, image: Image.Image, draw: ImageDraw, value: Any, min_val: float, max_val: float):
        if not value:
            return
            
        try:
            if isinstance(value, str):
                data = [float(x) for x in value.split(",")]
            else:
                data = [float(value)]
        except Exception:
            return

        x, y = self.rect["x"], self.rect["y"]
        w, h = self.rect["width"], self.rect["height"]
        
        color = self.config.get("color", "#00FF00")
        points = self.config.get("points", len(data))
        data = data[-points:]
        
        if len(data) < 2:
            return

        step_x = w / (len(data) - 1)
        range_val = max_val - min_val if max_val > min_val else 100

        line_points = []
        for i, val in enumerate(data):
            norm_val = (val - min_val) / range_val
            px = x + i * step_x
            py = y + h - (norm_val * h)
            line_points.append((px, py))
            
        draw.line(line_points, fill=color, width=self.config.get("thickness", 1))

    def info(self) -> dict:
        info = super().info()
        info["fields"].extend([
            {"name": "color", "value": "color"},
            {"name": "points", "value": "number"},
            {"name": "thickness", "value": "number"}
        ])
        return info

File: s1panel/widgets/line_chart.js
'use strict';
/*!
 * s1panel - widget/line_chart (native canvas, zero deps)
 * Copyright (c) 2024-2025 Tomasz Jaworski
 * GPL-3 Licensed
 */

function start_draw(context, rect) {
    context.save();
    context.beginPath();
    context.rect(rect.x, rect.y, rect.width, rect.height);
    context.clip();
}

function debug_rect(context, rect) {
    context.lineWidth = 1;
    context.strokeStyle = 'red';
    context.strokeRect(rect.x, rect.y, rect.width, rect.height);
}

function get_private(config) {
    if (!config._private) config._private = {};
    return config._private;
}

function draw(context, value, min, max, config) {

    return new Promise(fulfill => {

        const _private = get_private(config);
        const _rect    = config.rect;

        const _raw   = value.split(',').map(Number);
        const _count = Math.min(config.points || 60, _raw.length);
        const _data  = _raw.slice(_raw.length - _count);

        const _has_changed = (_private.last_value !== value);

        const _data_min = min ?? 0;
        const _data_max = max && max > 0 ? max : Math.max(..._data, 1);
        const _range    = _data_max - _data_min || 1;

        const scaleX = i  => _rect.x + (i / (_data.length - 1)) * _rect.width;
        const scaleY = v  => _rect.y + _rect.height - ((v - _data_min) / _range) * _rect.height;

        start_draw(context, _rect);

        // fill area beneath line
        if (config.area) {
            context.beginPath();
            _data.forEach((v, i) => {
                const x = scaleX(i), y = scaleY(v);
                i === 0 ? context.moveTo(x, y) : context.lineTo(x, y);
            });
            context.lineTo(scaleX(_data.length - 1), _rect.y + _rect.height);
            context.lineTo(scaleX(0), _rect.y + _rect.height);
            context.closePath();
            context.fillStyle = config.fill || 'rgba(0,212,255,0.15)';
            context.fill();
        }

        // draw the line
        context.beginPath();
        _data.forEach((v, i) => {
            const x = scaleX(i), y = scaleY(v);
            i === 0 ? context.moveTo(x, y) : context.lineTo(x, y);
        });
        context.strokeStyle = config.outline || '#00d4ff';
        context.lineWidth   = config.line_width || 1.5;
        context.lineJoin    = 'round';
        context.stroke();

        if (config.debug_frame) debug_rect(context, _rect);
        context.restore();

        if (_has_changed) _private.last_value = value;

        fulfill(_has_changed);
    });
}

function info() {
    return {
        name: 'line_chart',
        description: 'A smooth line/area chart (native canvas)',
        fields: [
            { name: 'outline',    value: 'color'   },
            { name: 'fill',       value: 'color'   },
            { name: 'points',     value: 'number'  },
            { name: 'area',       value: 'boolean' },
            { name: 'line_width', value: 'number'  },
        ]
    };
}

module.exports = { info, draw };